// graphene-pk11/build/cjs/token.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Token = exports.TokenFlag = void 0;
const tslib_1 = require("tslib");
const core = tslib_1.__importStar(require("./core"));
var TokenFlag;
(function (TokenFlag) {
    TokenFlag[TokenFlag["RNG"] = 1] = "RNG";
    TokenFlag[TokenFlag["WRITE_PROTECTED"] = 2] = "WRITE_PROTECTED";
    TokenFlag[TokenFlag["LOGIN_REQUIRED"] = 4] = "LOGIN_REQUIRED";
    TokenFlag[TokenFlag["USER_PIN_INITIALIZED"] = 8] = "USER_PIN_INITIALIZED";
    TokenFlag[TokenFlag["RESTORE_KEY_NOT_NEEDED"] = 32] = "RESTORE_KEY_NOT_NEEDED";
    TokenFlag[TokenFlag["CLOCK_ON_TOKEN"] = 64] = "CLOCK_ON_TOKEN";
    TokenFlag[TokenFlag["PROTECTED_AUTHENTICATION_PATH"] = 256] = "PROTECTED_AUTHENTICATION_PATH";
    TokenFlag[TokenFlag["DUAL_CRYPTO_OPERATIONS"] = 512] = "DUAL_CRYPTO_OPERATIONS";
    TokenFlag[TokenFlag["TOKEN_INITIALIZED"] = 1024] = "TOKEN_INITIALIZED";
    TokenFlag[TokenFlag["SECONDARY_AUTHENTICATION"] = 2048] = "SECONDARY_AUTHENTICATION";
    TokenFlag[TokenFlag["USER_PIN_COUNT_LOW"] = 65536] = "USER_PIN_COUNT_LOW";
    TokenFlag[TokenFlag["USER_PIN_FINAL_TRY"] = 131072] = "USER_PIN_FINAL_TRY";
    TokenFlag[TokenFlag["USER_PIN_LOCKED"] = 262144] = "USER_PIN_LOCKED";
    TokenFlag[TokenFlag["USER_PIN_TO_BE_CHANGED"] = 524288] = "USER_PIN_TO_BE_CHANGED";
    TokenFlag[TokenFlag["SO_PIN_COUNT_LOW"] = 1048576] = "SO_PIN_COUNT_LOW";
    TokenFlag[TokenFlag["SO_PIN_FINAL_TRY"] = 2097152] = "SO_PIN_FINAL_TRY";
    TokenFlag[TokenFlag["SO_PIN_LOCKED"] = 4194304] = "SO_PIN_LOCKED";
    TokenFlag[TokenFlag["SO_PIN_TO_BE_CHANGED"] = 8388608] = "SO_PIN_TO_BE_CHANGED";
    TokenFlag[TokenFlag["ERROR_STATE"] = 16777216] = "ERROR_STATE";
})(TokenFlag || (exports.TokenFlag = TokenFlag = {}));
class Token extends core.HandleObject {
    constructor(handle, lib) {
        super(handle, lib);
        this.getInfo();
    }
    getInfo() {
        const info = this.lib.C_GetTokenInfo(this.handle);
        this.label = core.removePadding(info.label);
        this.manufacturerID = core.removePadding(info.manufacturerID);
        this.model = core.removePadding(info.model);
        this.serialNumber = core.removePadding(Buffer.from(info.serialNumber).toString());
        this.flags = info.flags;
        this.maxSessionCount = info.maxSessionCount;
        this.sessionCount = info.sessionCount;
        this.maxRwSessionCount = info.maxRwSessionCount;
        this.rwSessionCount = info.rwSessionCount;
        this.maxPinLen = info.maxPinLen;
        this.minPinLen = info.minPinLen;
        this.totalPublicMemory = info.totalPublicMemory;
        this.freePublicMemory = info.freePublicMemory;
        this.totalPrivateMemory = info.totalPrivateMemory;
        this.freePrivateMemory = info.freePrivateMemory;
        this.hardwareVersion = info.hardwareVersion;
        this.firmwareVersion = info.firmwareVersion;
        this.utcTime = info.flags & TokenFlag.CLOCK_ON_TOKEN
            ? core.dateFromString(info.utcTime)
            : null;
    }
}
exports.Token = Token;
