// graphene-pk11/build/cjs/crypto/verify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Verify = void 0;
const tslib_1 = require("tslib");
const pkcs11 = tslib_1.__importStar(require("pkcs11js"));
const core = tslib_1.__importStar(require("../core"));
const mech_1 = require("../mech");
class Verify extends core.BaseObject {
    constructor(session, alg, key, lib) {
        super(lib);
        this.session = session;
        this.init(alg, key);
    }
    update(data) {
        try {
            const bytes = Buffer.from(data);
            this.lib.C_VerifyUpdate(this.session.handle, bytes);
        }
        catch (e) {
            try {
                this.final(Buffer.alloc(0));
            }
            catch (e) {
            }
            throw e;
        }
    }
    final(signature) {
        let res = false;
        try {
            res = this.lib.C_VerifyFinal(this.session.handle, signature);
        }
        catch (err) {
            if (!(err instanceof pkcs11.Pkcs11Error && err.code === pkcs11.CKR_SIGNATURE_INVALID)) {
                throw err;
            }
        }
        return res;
    }
    once(data, signature, cb) {
        const bytes = Buffer.from(data);
        if (cb) {
            this.lib.C_Verify(this.session.handle, bytes, signature, (err, data2) => {
                if (err instanceof pkcs11.Pkcs11Error && err.code === pkcs11.CKR_SIGNATURE_INVALID) {
                    cb(null, false);
                }
                else {
                    cb(err, data2);
                }
            });
        }
        else {
            let res = false;
            try {
                res = this.lib.C_Verify(this.session.handle, bytes, signature);
            }
            catch (err) {
                if (!(err instanceof pkcs11.Pkcs11Error && err.code === pkcs11.CKR_SIGNATURE_INVALID)) {
                    throw err;
                }
            }
            return res;
        }
    }
    init(alg, key) {
        const pMech = mech_1.Mechanism.create(alg);
        this.lib.C_VerifyInit(this.session.handle, pMech, key.handle);
    }
}
exports.Verify = Verify;
