// graphene-pk11/build/cjs/core/type.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isString = isString;
exports.isNumber = isNumber;
exports.isBoolean = isBoolean;
exports.isUndefined = isUndefined;
exports.isNull = isNull;
exports.isEmpty = isEmpty;
exports.isFunction = isFunction;
exports.isObject = isObject;
exports.isArray = isArray;
exports.isFlag = isFlag;
exports.removePadding = removePadding;
function isString(v) {
    return typeof (v) === "string";
}
function isNumber(v) {
    return typeof (v) === "number";
}
function isBoolean(v) {
    return typeof (v) === "boolean";
}
function isUndefined(v) {
    return typeof (v) === "undefined";
}
function isNull(v) {
    return v === null;
}
function isEmpty(v) {
    return isUndefined(v) || isNull(v);
}
function isFunction(v) {
    return typeof (v) === "function";
}
function isObject(v) {
    return typeof (v) === "object";
}
function isArray(v) {
    return Array.isArray(v);
}
function isFlag(v, fv) {
    return (v & fv) === fv;
}
function removePadding(text) {
    return text.replace(/\0.*/g, "").trim();
}
