// graphene-pk11/build/es2015/crypto/digest.js
import * as core from "../core";
import * as mech from "../mech";
export class Digest extends core.BaseObject {
    constructor(session, alg, lib) {
        super(lib);
        this.session = session;
        this.init(alg);
    }
    update(data) {
        try {
            const bytes = Buffer.from(data);
            this.lib.C_DigestUpdate(this.session.handle, bytes);
        }
        catch (e) {
            try {
                this.final();
            }
            catch (e) {
            }
            throw e;
        }
    }
    final() {
        const digest = Buffer.alloc(1024);
        const res = this.lib.C_DigestFinal(this.session.handle, digest);
        return res;
    }
    once(data, cb) {
        const digest = Buffer.alloc(1024);
        const bytes = Buffer.from(data);
        if (cb) {
            this.lib.C_Digest(this.session.handle, bytes, digest, cb);
        }
        else {
            return this.lib.C_Digest(this.session.handle, bytes, digest);
        }
    }
    init(alg) {
        const pMech = mech.Mechanism.create(alg);
        this.lib.C_DigestInit(this.session.handle, pMech);
    }
}
