// graphene-pk11/build/es2015/objects/certs/x509.js
import { __decorate } from "tslib";
import { attribute } from "../../core";
import { Certificate } from "./cert";
export var JavaMIDP;
(function (JavaMIDP) {
    JavaMIDP[JavaMIDP["Unspecified"] = 0] = "Unspecified";
    JavaMIDP[JavaMIDP["Manufacturer"] = 1] = "Manufacturer";
    JavaMIDP[JavaMIDP["Operator"] = 2] = "Operator";
    JavaMIDP[JavaMIDP["ThirdParty"] = 3] = "ThirdParty";
})(JavaMIDP || (JavaMIDP = {}));
export class X509Certificate extends Certificate {
    get serialNumber() {
        return this.get("serial").toString("hex");
    }
    set serialNumber(v) {
        this.set("serial", Buffer.from(v, "hex"));
    }
}
__decorate([
    attribute("subject")
], X509Certificate.prototype, "subject", void 0);
__decorate([
    attribute("id")
], X509Certificate.prototype, "id", void 0);
__decorate([
    attribute("issuer")
], X509Certificate.prototype, "issuer", void 0);
__decorate([
    attribute("value")
], X509Certificate.prototype, "value", void 0);
__decorate([
    attribute("url")
], X509Certificate.prototype, "url", void 0);
__decorate([
    attribute("ski")
], X509Certificate.prototype, "subjectKeyIdentifier", void 0);
__decorate([
    attribute("aki")
], X509Certificate.prototype, "authorityKeyIdentifier", void 0);
__decorate([
    attribute("javaDomain")
], X509Certificate.prototype, "java", void 0);
