// graphene-pk11/build/cjs/objects/certs/cert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Certificate = exports.CertificateCategory = exports.CertificateType = void 0;
const tslib_1 = require("tslib");
const pkcs11 = tslib_1.__importStar(require("pkcs11js"));
const core_1 = require("../../core");
const storage_1 = require("../storage");
var CertificateType;
(function (CertificateType) {
    CertificateType[CertificateType["X_509"] = pkcs11.CKC_X_509] = "X_509";
    CertificateType[CertificateType["X_509_ATTR_CERT"] = pkcs11.CKC_X_509_ATTR_CERT] = "X_509_ATTR_CERT";
    CertificateType[CertificateType["WTLS"] = pkcs11.CKC_WTLS] = "WTLS";
})(CertificateType || (exports.CertificateType = CertificateType = {}));
var CertificateCategory;
(function (CertificateCategory) {
    CertificateCategory[CertificateCategory["Unspecified"] = 0] = "Unspecified";
    CertificateCategory[CertificateCategory["TokenUser"] = 1] = "TokenUser";
    CertificateCategory[CertificateCategory["Authority"] = 2] = "Authority";
    CertificateCategory[CertificateCategory["OtherEntity"] = 3] = "OtherEntity";
})(CertificateCategory || (exports.CertificateCategory = CertificateCategory = {}));
class Certificate extends storage_1.Storage {
}
exports.Certificate = Certificate;
tslib_1.__decorate([
    (0, core_1.attribute)("certType")
], Certificate.prototype, "type", void 0);
tslib_1.__decorate([
    (0, core_1.attribute)("trusted")
], Certificate.prototype, "trusted", void 0);
tslib_1.__decorate([
    (0, core_1.attribute)("certCategory")
], Certificate.prototype, "category", void 0);
tslib_1.__decorate([
    (0, core_1.attribute)("checkValue")
], Certificate.prototype, "checkValue", void 0);
tslib_1.__decorate([
    (0, core_1.attribute)("startDate")
], Certificate.prototype, "startDate", void 0);
tslib_1.__decorate([
    (0, core_1.attribute)("endDate")
], Certificate.prototype, "endDate", void 0);
